cmd/gen-key/ecdsa.go

```diff
@@ -0,0 +1,195 @@
+			}
+			gotCurve = true
+		case pkcs11.CKA_EC_POINT:
+			x, y := elliptic.Unmarshal(curve, a.Value)
+			if x == nil {
+				// http://docs.oasis-open.org/pkcs11/pkcs11-curr/v2.40/os/pkcs11-curr-v2.40-os.html#_ftn1
+				// PKCS#11 v2.20 specified that the CKA_EC_POINT was to be stored in a DER-encoded
+				// OCTET STRING.
+				var point asn1.RawValue
+				_, err = asn1.Unmarshal(a.Value, &point)
+				if err != nil {
+					return nil, fmt.Errorf("Failed to unmarshal returned CKA_EC_POINT: %s", err)
+				}
+				if len(point.Bytes) == 0 {
+					return nil, errors.New("Invalid CKA_EC_POINT value returned, OCTET string is empty")
+				}
+				x, y = elliptic.Unmarshal(curve, point.Bytes)
+				if x == nil {
+					fmt.Println(point.Bytes)
+					return nil, errors.New("Invalid CKA_EC_POINT value returned, point is malformed")
+				}
+			}
+			pubKey.X, pubKey.Y = x, y
+			gotPoint = true
+			log.Printf("\tX: %X\n", pubKey.X.Bytes())
+			log.Printf("\tY: %X\n", pubKey.Y.Bytes())
+		}
+	}
+	if !gotPoint || !gotCurve {
+		return nil, errors.New("Couldn't retrieve EC point or EC parameters")
+	}
+	return pubKey, nil
+}
+
+var curveToHash = map[elliptic.Curve]crypto.Hash{
+	elliptic.P224(): crypto.SHA256,
+	elliptic.P256(): crypto.SHA256,
+	elliptic.P384(): crypto.SHA384,
+	elliptic.P521(): crypto.SHA512,
+}
+
+var hashToString = map[crypto.Hash]string{
+	crypto.SHA256: "SHA-256",
+	crypto.SHA384: "SHA-384",
+	crypto.SHA512: "SHA-512",
+}
+
+func ecVerify(ctx Ctx, session pkcs11.SessionHandle, object pkcs11.ObjectHandle, pub *ecdsa.PublicKey) error {
+	err := ctx.SignInit(session, []*pkcs11.Mechanism{pkcs11.NewMechanism(pkcs11.CKM_ECDSA, nil)}, object)
+	if err != nil {
+		return fmt.Errorf("failed to initialize signing operation: %s", err)
+	}
+	msg, err := getRandomBytes(ctx, session)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve random data: %s", err)
+	}
+	log.Printf("\tConstructed random number: %d (%X)\n", big.NewInt(0).SetBytes(msg), msg)
+	hashFunc := curveToHash[pub.Curve].New()
+	hashFunc.Write(msg)
+	hash := hashFunc.Sum(nil)
+	log.Printf("\tMessage %s hash: %X\n", hashToString[curveToHash[pub.Curve]], hash)
+	signature, err := ctx.Sign(session, hash)
+	if err != nil {
+		return fmt.Errorf("failed to sign data: %s", err)
+	}
+	log.Printf("\tMessage signature: %X\n", signature)
+	r := big.NewInt(0).SetBytes(signature[:len(signature)/2])
+	s := big.NewInt(0).SetBytes(signature[len(signature)/2:])
+	if !ecdsa.Verify(pub, hash[:], r, s) {
+		return errors.New("failed to verify ECDSA signature over test data")
+	}
+	log.Println("\tSignature verified")
+	return nil
+}
+
+func ecGenerate(ctx Ctx, session pkcs11.SessionHandle, label, curveStr string, compatMode bool) (*ecdsa.PublicKey, error) {
+	curve, present := stringToCurve[curveStr]
+	if !present {
+		return nil, errors.New("curve not supported")
+	}
+	log.Printf("Generating ECDSA key with curve %s\n", curveStr)
+	m, pubTmpl, privTmpl := ecArgs(label, curve, compatMode)
```

Comment:
Let's call this "mechanism" instead of m, since it's not immediately clear from context what it is. As a general rule, I think single-letter names are best for loop variables and (sometimes) the name of the `this` pointer in methods.